Publish sitemap for doc pages

bjfish opened the pull request:
Related to #190 

I tested the publish_sitemap method with my own S3.

I don't know how to test the publish / revert methods locally yet (which now call publish_sitemap).  Should I test these? If so, how would I point a hex client at my localhost?

bjfish commented on web/controllers/api/docs_controller.ex:
```diff
@@ -199,4 +201,29 @@ defmodule HexWeb.API.DocsController do
+ xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd">
+ <%= for package <- packages do %><url>
+     <loc>http://hexdocs.pm/<%= package.name %></loc>
+     <lastmod><%=  Ecto.DateTime.to_iso8601(package.updated_at) %></lastmod>
```
I don't think this is absolutely accurate for lastmod. Release updated_at might be better.

ericmj replied:
I think we need to add a new datetime field to package that is set to when the main (non-versioned) docs were last updated.

The reason for this is because in the future we will add some logic that will conditionally update the main docs. For example pre-releases should not automatically replace main docs.

ericmj commented on web/models/package.ex:
```diff
@@ -131,6 +131,12 @@ defmodule HexWeb.Package do
+    from( p in Package,
+         join: r in assoc(p, :releases),
+         where: r.has_docs == true)
+  end
```
This will generate some duplicate package names since a package can have multiple releases.

ericmj commented on web/controllers/api/docs_controller.ex:
```diff
@@ -28,6 +28,7 @@ defmodule HexWeb.API.DocsController do
 
   def delete(conn, _params) do
     revert(conn.assigns.release)
+    publish_sitemap()
```
This function should run inside the async task, i.e. inside the `revert/1` function.

ericmj commented on web/controllers/api/docs_controller.ex:
```diff
@@ -38,6 +39,7 @@ defmodule HexWeb.API.DocsController do
     case parse_tarball(package, release, user, body) do
       :ok ->
         location = HexWeb.Utils.docs_tarball_url(package, release)
+        publish_sitemap()
```
This should run inside `upload_docs/4`.

bjfish commented:
I've fixed the `all_has_docs` query and moved the `publish_sitemap()` calls to the functions you mentioned. 

Should adding a new datetime field to package be a part of this PR?

ericmj commented on web/controllers/api/docs_controller.ex:
```diff
@@ -194,9 +196,38 @@ defmodule HexWeb.API.DocsController do
+        ~s'  <% end %>' <>
+        ~s'</urlset>'
+
+      result = EEx.eval_string template, packages: packages
```
We should use `EEx.compile_string` or use phoenix templates because `eval_string` will have bad performance.

ericmj commented on web/controllers/api/docs_controller.ex:
```diff
@@ -194,9 +196,38 @@ defmodule HexWeb.API.DocsController do
+      result = EEx.eval_string template, packages: packages
+
+      # publish sitemap
+      HexWeb.Store.put_docs_page("sitemap.xml", "docspage/sitemap.xml", result)
```
Too much indentation.

ericmj commented:
> Should adding a new datetime field to package be a part of this PR?

If you don't mind yes, otherwise I can do it.

lexmag commented on web/controllers/api/docs_controller.ex:
```diff
@@ -194,9 +196,40 @@ defmodule HexWeb.API.DocsController do
+    HexWeb.Store.put_docs_page("sitemap.xml", "docspage/sitemap.xml", result)
+  end
+
+  @sitemap_template "" <>
```
It'd be better to use heredoc definition:
```elixir
@sitemap_template """
<?xml version="1.0" encoding="utf-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
...
</urlset>
"""
```

lexmag commented on web/controllers/api/docs_controller.ex:
```diff
@@ -194,9 +196,40 @@ defmodule HexWeb.API.DocsController do
+    ~s'</urlset>'
+  require EEx
+  EEx.function_from_string :defp, :render_sitemap, @sitemap_template, [:packages]
+
```
Needless line. :scissors:

lexmag commented on web/models/package.ex:
```diff
@@ -131,6 +131,14 @@ defmodule HexWeb.Package do
   end
 
+  def all_has_docs do
+    from( p in Package,
```
There is one extra space before `p`. :scissors:

bjfish commented:
Thanks @lexmag , I've updated per your feedback.

@ericmj I've added a new datetime to packages. One thing left to do is updating this new date when the root docs are updated.

lexmag commented on web/controllers/api/docs_controller.ex:
```diff
@@ -194,9 +196,40 @@ defmodule HexWeb.API.DocsController do
+    </urlset>
+  """
+  require EEx
+  EEx.function_from_string :defp, :render_sitemap, String.lstrip(@sitemap_template), [:packages]
```
`String.lstrip` has no effect on `@sitemap_template`, we can just remove it.

bjfish replied:
I had to lstrip because the XML spec doesn't allow whitespace before the leading xml declaration.

bjfish replied:
I tried this without the lstrip and it appears to be stripping out the whitespace so I've removed the lstrip.

lexmag commented on web/controllers/api/docs_controller.ex:
```diff
@@ -194,9 +196,40 @@ defmodule HexWeb.API.DocsController do
+
+  @sitemap_template """
+  <?xml version="1.0" encoding="utf-8"?>
+    <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
```
Maybe `<urlset` should have the same indentation with `<?xml`?

bjfish replied:
I've fixed this indentation.

lexmag commented on priv/repo/migrations/20160227170838_add_docs_root_updated_at_to_packages.exs:
```diff
@@ -0,0 +1,23 @@
+      remove :docs_root_updated_at
+    end
+  end
+
```
Needless line. :scissors:

bjfish replied:
I've removed this line.

ericmj commented on web/models/package.ex:
```diff
@@ -9,6 +9,7 @@ defmodule HexWeb.Package do
   schema "packages" do
     field :name, :string
     field :meta, :map
+    field :docs_root_updated_at, Ecto.DateTime
```
Should call this just `docs_updated_at`.

bjfish commented:
@ericmj I've updated the new field name to `docs_updated_at`

ericmj commented:
@bjfish Thanks!
